File: src/VRE.Vridge.API.Client/Proxy/APIClient.cs
using System;
using System.Collections.Generic;
using NetMQ.Sockets;
using VRE.Vridge.API.Client.Helpers;
using VRE.Vridge.API.Client.Messages;
using VRE.Vridge.API.Client.Messages.Control;
using VRE.Vridge.API.Client.Messages.Control.Requests;
using VRE.Vridge.API.Client.Messages.Control.Responses;
using VRE.Vridge.API.Client.Proxy.Broadcasts;
using VRE.Vridge.API.Client.Proxy.Controller;
using VRE.Vridge.API.Client.Proxy.HeadTracking;

namespace VRE.Vridge.API.Client.Proxy
{
    public class APIClient
    {
        public enum Endpoints
        {
            HeadTracking,
            Controller,
            Broadcast
        }

        private RequestSocket     controlSocket;        
        
        private readonly string serverAddress = "localhost";
        private readonly string appName = "default";

        public APIClient(string appName)
        {
            this.appName = appName;            
        }

        public APIClient(string ip, string appName)
        {
            serverAddress = ip;
            this.appName = appName;            
        }

        /// <summary>
        /// Sends control request to see what APIs are available. 
        /// May return null if control connection dies (automatic reconnect will follow). 
        /// </summary>        
        public APIStatus GetStatus()
        {
            APIStatus status;

            ConnectToControlSocket();

            controlSocket.SendAsJson(new ControlRequestHeader(ControlRequestCode.RequestStatus));            
            var success = controlSocket.TryReceiveJson(out status, 500);

            controlSocket.Close();

            if (success)
            {
                return status;
            }            

            return null;
        }

        public T CreateProxy<T>(bool keepAlive = true)
        {            
            // Find out which endpoint name should be used
            string endpointName;
            if (typeof(T) == typeof(HeadTrackingProxy))
            {
                endpointName = EndpointNames.HeadTracking;
            }
            else if (typeof(T) == typeof(ControllerProxy))
            {
                endpointName = EndpointNames.Controller;
            }
            else if (typeof(T) == typeof(BroadcastProxy))
            {
                endpointName = EndpointNames.Broadcast;
            }
            else
            {
                throw new ArgumentException("Invalid proxy requested.");
            }

            ConnectToControlSocket();

            // Try requesting given endpoint
            bool success = controlSocket.TrySendAsJson(new RequestEndpoint(endpointName, appName), 1000);

            if (!success)
            {
                HandleControlConnectionException(new Exception("API server timeout."));
            }
                

            EndpointCreated response;
            success = controlSocket.TryReceiveJson(out response, 1000);

            controlSocket.Close();

            if (!success)
            {
                HandleControlConnectionException(new Exception("API server timeout."));
            }

            if (response.Code == (int) ControlResponseCode.InUse)
            {
                HandleControlConnectionException(new Exception("API already in use by another client."));
            }

            // Initialize the proxy of requested type
            var connectionString = $"tcp://{serverAddress}:{response.Port}";
            if (typeof(T) == typeof(HeadTrackingProxy))
            {
                return (T) Convert.ChangeType(new HeadTrackingProxy(connectionString, keepAlive), typeof(T));
            }
            if (typeof(T) == typeof(ControllerProxy))
            {
                return (T) Convert.ChangeType(new ControllerProxy(connectionString, keepAlive), typeof(T));
            }
            if (typeof(T) == typeof(BroadcastProxy))
            {
                return (T) Convert.ChangeType(new BroadcastProxy(connectionString), typeof(T));
            }

            throw new ArgumentException("Invalid proxy requested.");
        }

        private void ConnectToControlSocket()
        {
            controlSocket = new RequestSocket();            
            controlSocket.Connect(ControlEndpoint);
        }

        protected void HandleControlConnectionException(Exception x)
        {
            // Reset socket state to continue accepting new requests*            
            controlSocket.Close();
            controlSocket = new RequestSocket();
            controlSocket.Connect(ControlEndpoint);

            // Let client know that something went wrong
            throw x;

            /* *This can be also done with ZMQ_REQ_RELAXED + ZMQ_REQ_CORRELATE
               options instead of brute-force-like restarting the socket    
               but NetMQ does not seem to support these options */
        }

        private string ControlEndpoint => $"tcp://{serverAddress}:38219";
    }
}

File: src/java/VRE.Vridge.API/VRE.Vridge.API.Client/src/main/java/com/riftcat/vridge/api/client/java/APIClient.java
package com.riftcat.vridge.api.client.java;

import com.riftcat.vridge.api.client.java.control.ControlRequestCode;
import com.riftcat.vridge.api.client.java.control.ControlResponseCode;
import com.riftcat.vridge.api.client.java.control.requests.ControlRequestHeader;
import com.riftcat.vridge.api.client.java.control.requests.RequestEndpoint;
import com.riftcat.vridge.api.client.java.control.responses.APIStatus;
import com.riftcat.vridge.api.client.java.control.responses.EndpointCreated;
import com.riftcat.vridge.api.client.java.proxy.BroadcastProxy;
import com.riftcat.vridge.api.client.java.proxy.ClientProxyBase;
import com.riftcat.vridge.api.client.java.proxy.ControllerProxy;
import com.riftcat.vridge.api.client.java.proxy.HeadTrackingProxy;
import com.riftcat.vridge.api.client.java.proxy.VRidgeApiProxy;
import com.riftcat.vridge.api.client.java.utils.ILog;
import com.riftcat.vridge.api.client.java.utils.SocketHelpers;

import org.zeromq.ZContext;
import org.zeromq.ZMQ;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.concurrent.TimeoutException;

public class APIClient {

    public final static int HEADTRACKING = 0;
    public final static int CONTROLLER = 1;
    public final static int BROADCASTS = 2;

    public static ZContext ZContext;

    private HashMap<Integer, VRidgeApiProxy> proxies;
    private String serverAddress = "tcp://localhost";

    // Connections with same app name will not result in "endpoint in use" response
    private String appName = "";

    public APIClient(String appName){
        ZContext = new ZContext(4);
        proxies = new HashMap<Integer, VRidgeApiProxy>();

        this.appName = appName;
    }

    public APIClient(String ip, String appName){
        this(appName);
        serverAddress = ip;
    }

    /// <summary>
    /// Sends control request to see what APIs are available.
    /// May return null if control connection dies (automatic reconnect will follow).
    /// </summary>
    public APIStatus GetStatus() throws Exception {

        ZMQ.Socket controlSocket = createControlSocket();
        if (controlSocket == null)
        {
            return null;
        }

        SocketHelpers.SendAsJson(controlSocket, new ControlRequestHeader(ControlRequestCode.RequestStatus));
        APIStatus status = SocketHelpers.ReceiveByJson(controlSocket, APIStatus.class);
        APIClient.ZContext.destroySocket(controlSocket);

        if (status != null){
            return status;
        }

        throw new Exception("Could not read API status.");
    }

    public <T extends VRidgeApiProxy> T getProxy(int proxyType) throws Exception {

        VRidgeApiProxy proxy = proxies.get(proxyType);
        ZMQ.Socket controlSocket = createControlSocket();

        if(proxy == null){

            String endpointName = null;
            switch (proxyType)
            {
                case HEADTRACKING:
                    endpointName = EndpointNames.HeadTracking;
                    break;
                case CONTROLLER:
                    endpointName = EndpointNames.Controller;
                    break;
                case BROADCASTS:
                    endpointName = EndpointNames.Broadcast;
                    break;
            }

            if(endpointName == null){{
                throw new IllegalArgumentException("Invalid proxy type was requested.");
            }}

            SocketHelpers.SendAsJson(controlSocket, new RequestEndpoint(endpointName, appName));
            EndpointCreated response = SocketHelpers.ReceiveByJson(controlSocket, EndpointCreated.class);
            APIClient.ZContext.destroySocket(controlSocket);

            if(response == null ){
                throw new TimeoutException("API server timeout");
            }

            if(response.Code == ControlResponseCode.InUse){
                throw new Exception("API endpoint in use.");
            }

            switch (proxyType){
                case HEADTRACKING:
                    proxies.put(proxyType, new HeadTrackingProxy("tcp://" + serverAddress + ":" + response.Port, true));
                    break;
                case CONTROLLER:
                    proxies.put(proxyType, new ControllerProxy("tcp://" + serverAddress + ":" + response.Port));
                    break;
                case BROADCASTS:
                    proxies.put(proxyType, new BroadcastProxy("tcp://" + serverAddress + ":" + response.Port));
                    break;
            }
        }

        return (T) proxies.get(proxyType);
    }

    public void disconnectProxy(int proxyType)
    {
        VRidgeApiProxy proxy = proxies.get(proxyType);

        if (proxy == null) return;

        proxy.disconnect();
        proxies.put(proxyType, null);
    }

    public void disconnectAll() {
        for(int proxyId : proxies.keySet()){
            disconnectProxy(proxyId);
        }
    }

    private String getEndpointAddress() {

        return "tcp://" + serverAddress + ":38219";
    }

    private ZMQ.Socket createControlSocket(){
        String ctrlAddress = getEndpointAddress();

        ZMQ.Socket controlSocket = ZContext.createSocket(ZMQ.REQ);
        controlSocket.connect(ctrlAddress);
        controlSocket.setSendTimeOut(1000);
        controlSocket.setReceiveTimeOut(1000);
        return controlSocket;
    }
}
